### PR TITLE
Wpta 622

### PR DESCRIPTION
**Pull request by ntashev**

### CODE REVIEW

File: pinger.go [hunk 1]
```diff
@@ -14,37 +14,69 @@ import (
...
+func GetSettings() (int, map[string]string, time.Time,  error) {
 	target := os.Getenv("TARGET_URL")
 	method := os.Getenv("METHOD")
-	interval, err := strconv.Atoi(os.Getenv("INTERVAL"))
+	interval, err1 := strconv.Atoi(os.Getenv("INTERVAL"))
+
+	timeToRun := time.Time{}
+	var err2 error
+	if os.Getenv("TIME") != "" {
+		timeToRun, err2 = time.Parse("15:04:05", os.Getenv("TIME"))
```
Comment (philip-ince) [inline pinger.go]: whats this time for?

File: pinger.go [hunk 2]
```diff
@@ -14,37 +14,69 @@ import (
...
+			}
 		}
-		log.Printf("Sleeping for %d seconds", interval)
-		time.Sleep(time.Duration(interval) * time.Second)
 	}
 }
 
+//GetDurationToWait Get duration till next run
+func GetDurationToWait(interval int, timeToRun time.Time) time.Duration {
+	now := time.Now().UTC()
```
Comment (philip-ince) [inline pinger.go]: maybe i'm missing this but where does it cope with time zone?

File: pinger.go (hunk 1 above)
Comment (ntashev) [inline pinger.go]: For reconciliation we want to run the ping in a particular time of the day. if you are asking about "15:04:05", this is how the parsing format is defined in Go. This is equivalent to "HH:mm:ss".

File: pinger.go (hunk 2 above)
Comment (ntashev) [inline pinger.go]: I made the pinger work only in UTC. The time zone difference will be added in the yml that configures the pinger.

File: pinger.go [hunk 3]
```diff
@@ -14,37 +14,69 @@ import (
...
+			if err != nil {
+				log.Print(err)
+			}
 		}
-		log.Printf("Sleeping for %d seconds", interval)
-		time.Sleep(time.Duration(interval) * time.Second)
 	}
 }
 
+//GetDurationToWait Get duration till next run
```
Comment (johnpeterharvey) [inline pinger.go]: space after //

File: pinger.go [hunk 4]
```diff
@@ -14,37 +14,69 @@ import (
...
+func GetDurationToWait(interval int, timeToRun time.Time) time.Duration {
+	now := time.Now().UTC()
+	nextRun := time.Date(now.Year(), now.Month(), now.Day(), timeToRun.Hour(), timeToRun.Minute(), timeToRun.Second(), 0, time.UTC)
+
+	if !nextRun.After(now) {
```
Comment (johnpeterharvey) [inline pinger.go]: why have this logic?

File: pinger.go [hunk 5]
```diff
@@ -14,37 +14,69 @@ import (
...
+			err := DoCall(client, settings)
+			if err != nil {
+				log.Print(err)
+			}
+			log.Printf("Sleeping for %d seconds", interval)
+			time.Sleep(time.Duration(interval) * time.Second)
+		}
+	} else {
+		for {
+			duration := GetDurationToWait(interval, timeToRun)
```
Comment (johnpeterharvey) [inline pinger.go]: what if it's 16:05, and then time to run is 16:00 - then will duration be -5? it should have a negative check and not run if it's due to run in the past surely?

**Comment by johnpeterharvey** [comment]
This needs test updates as well

File: pinger.go [hunk 6]
```diff
@@ -14,37 +14,69 @@ import (
...
 		os.Exit(1)
 	}
 
 	client := &http.Client{}
 
-	for {
-		err := DoCall(client, settings)
-		if err != nil {
-			log.Print(err)
+	if time.Time.IsZero(timeToRun) {
```
Comment (johnpeterharvey) [inline pinger.go]: maybe put a comment here to explain what the different logic branches are for

File: pinger.go (hunk 5 above)
Comment (ntashev) [inline pinger.go]: We have the check above and if the time to run is in the past we add the configured duration. Have a look at your previous comment.

**Comment by philip-ince** [comment]
LGTM

**Pull request merged**